### DISCUSSION TITLE
Internationalization and Translations

### DISCUSSION THREAD

**Discussion by BravoBaldo** [harvard-edge/cs249r_book#347] [Q&A (General)]
Hello, 
my name is Baldassarre Cesarano. You can find me on GitHub at https://github.com/BravoBaldo.
I'm interested in Italian translations.

In Sphinx firstly convert source files in "*.po" format
then I use OmegaT to translate these .po files
and a tool (sphinx-build) replace dinamically translated phrases WITHOUT TOUCH the source files.

I read article https://www.appsilon.com/post/shiny-i18n-quarto-multilingual-text but it seems that it requires modifying the sources.
Is there a procedure defined to translate cs249r_book?

Any ideas?

**Comment by profvjreddi** [comment]
Hi @BravoBaldo 👋 

Translation and internationalization have been on the to-do list for a little while, but I am glad you brought this up! 

I was originally looking into some online automatic translation sites since that would be a lot easier to manage and maintain than manually converting the source files one by one. Would you happen to have any experience or thoughts about this approach?

**Reply by SergeStinckwich** [reply]
> Hi @BravoBaldo 👋
> 
> Translation and internationalization have been on the to-do list for a little while, but I am glad you brought this up!
> 
> I was originally looking into some online automatic translation sites since that would be a lot easier to manage and maintain than manually converting the source files one by one. Would you happen to have any experience or thoughts about this approach?


Dear all, I would be interested by translations also. I have done that for several books already from English to French.
I understand that you are looking into some way to automatically translate the book, but I think this is also important to have quality translations. I guess the solutions would be to combine both.

**Comment by BravoBaldo** [comment]
Ciao everyone,
I found my way and I'll tell you how to do it.
 
A little premise,
- I'm not a professional translator, I've learned a little English lately!
- Given my age and the need/desire to learn I need to read a lot of documentation in a short time and, above all, maintain my attention! ...So I translate a lot!
- Google Translate has improved a lot but every sentence must at least be re-read to maintain the context and terminology.
- A little attention to the tables.
- As a tool I use OmegaT + Okapi Filter (for md/qmd files).
 
This is my procedure:
1) In OmegaT copy all contents\*.qmd files (or link to them) renaming to *_it.qmd + "_quarto.yml" & "index.qmd"
2) Translate 
3) copy (or link) all *.qmd files translated NEAR each original (the names are different then, for example, you will have:
     ...
     ...\contents\acknowledgements\acknowledgements.qmd    (original)
     ...\contents\acknowledgements\acknowledgements_it.qmd (translated)
...
This allow to NOT change original files.
The only files I have to change are 
_quarto.yml and index.qmd (it seems that it is not possible to rename them, but I still need to investigate)

 

On
https://github.com/BravoBaldo/Translations/tree/main/TinyML/cs249r_book/_book_it
I have "published" (just for you), a partial translation (the first 8 chapters, I expect to finish in the next month or 2...)
 
Let me know what you think
 
Baldo

**Reply by profvjreddi** [reply]
Thanks this is super helpful! 

It is good to learn about tools like OmegaT, I haven't used it myself. 

I was exploring Lokalise / Transifex. Have you used this by any chance?

> On https://github.com/BravoBaldo/Translations/tree/main/TinyML/cs249r_book/_book_it I have "published" (just for you), a partial translation (the first 8 chapters, I expect to finish in the next month or 2...)

Awesome, thanks for sharing. 

Question: How would this be integrated into the website or the PDF build? Do we have to update the quarto yml file to have different versions, sorta like here: https://quarto.org/docs/authoring/language.html

**Reply by BravoBaldo** [reply]
The good technique (in my opinion) is the one used by Sphinx, that is:

1. Leave the original files unchanged
2. From the source files you get the sentences used as indexes of a "dictionary" forming .po files, placing them in a separate directory for each language.
     **Note**: The .po files are managed by OmegaT
4. During rendering you indicate the language and each sentence is changed on the fly with the translated one for any output format (pdf, word, html...).

https://projects.blender.org/blender/blender-manual-translations
A big project that more or less uses the same concept.

Sphinx documentation itself is managed with transifex
https://github.com/sphinx-doc/sphinx
https://app.transifex.com/sphinx-doc/sphinx-doc/language/it/

Yes, I use transifex, but I prefer not to use their interface to translate the "sentences", I directly provide them with the translated .po files on my local PC.
This is because, internally, OmegaT also breaks down the text into sentences and I keep MY OWN database of translated sentences (from all books/text)  ;)

As for Quarto's internationalization methodology, I need to investigate.
We'll probably have to ask them for support.

Baldo

**Comment by BravoBaldo** [comment]
This:
[https://docs.ropensci.org/babeldown/articles/quarto.html](https://docs.ropensci.org/babeldown/articles/quarto.html) (and the [example](https://devdevguide.netlify.app/))
seems to be a good solution.

It also translates the pdf!

What do you think?

Baldo

**Comment by profvjreddi** [comment]
The babel thing looks interesting. Thanks for sharing it. The only way we'll know which translation to pick is to try some of these out and figure out which one is the best option for long-term sustainability. 

I am tempted to use something like https://lokalise.com/ because it is easier to maintain and has GitHub integrations. We could create an automatic workflow to build the translations on the fly, which might be easier for constantly evolving content. 

That said, we should try both approaches to see what makes the most sense.

Do you want to try out the Babel one and I can try the Lokalise or get someone to help me with it?

**Comment by iskode** [comment]
Hi all @profvjreddi @SergeStinckwich @BravoBaldo,
I'd like to warm-up the discussion as I am interested in  the French translation.
I want to pick up a tool test it and share my experiences so we can settle on
a tool to move forward.
I want to test [lokalise]( https://lokalise.com/ ). Is anyone testing currently testing it so as to avoid duplicate work ?

Ismael

**Comment by BravoBaldo** [comment]
Hi Ismael,
I use OmegaT (https://omegat.org/)
 
Baldo

> 
P.S.: Rather than the translation itself, I'm more interested in the fact that it forces me to read and therefore study ;).
The translation is a side effect!

**Reply by iskode** [reply]
Hi Bravo,
thank you so much for letting me know.
That's great !
